shell/platform/linux/fl_view.cc

```diff
@@ -674,57 +628,58 @@ void fl_view_add_gl_area(FlView* view,
-                        GdkRectangle* geometry) {
-  gtk_widget_show(widget);
-  fl_view_add_pending_child(view, widget, geometry);
-}
-
-GList* find_child(GList* list, GtkWidget* widget) {
-  for (GList* i = list; i; i = i->next) {
-    FlViewChild* child = reinterpret_cast<FlViewChild*>(i->data);
-    if (child && child->widget == widget)
-      return i;
+                        GdkRectangle* geometry,
+                        GPtrArray* mutations) {
+  FlClippingView* clipping_view;
+  if (view->last_clipping_views->count(widget)) {
+    clipping_view = view->last_clipping_views->at(widget);
+    view->last_clipping_views->erase(widget);
+  } else {
+    clipping_view = FL_CLIPPING_VIEW(fl_clipping_view_new());
   }
-  return nullptr;
+  view->current_clipping_views->insert({widget, clipping_view});
+  fl_clipping_view_reset(clipping_view, widget, geometry, mutations);
+
+  gtk_widget_show_all(GTK_WIDGET(clipping_view));
+  fl_view_add_pending_child(view, GTK_WIDGET(clipping_view));
 }
 
-void fl_view_end_frame(FlView* view) {
-  for (GList* pending_child = view->pending_children_list; pending_child;
-       pending_child = pending_child->next) {
-    FlViewChild* pending_view_child =
-        reinterpret_cast<FlViewChild*>(pending_child->data);
-    GList* child = find_child(view->children_list, pending_view_child->widget);
+void fl_view_end_frame(FlView* self) {
+  for (GtkWidget* pending_child : *self->pending_children_list) {
+    auto child = std::find(self->children_list->begin(),
+                           self->children_list->end(), pending_child);
 
-    if (child) {
+    if (child != self->children_list->end()) {
       // existing child
-      g_free(child->data);
-      child->data = nullptr;
+      *child = nullptr;
     } else {
       // newly added child
-      gtk_widget_set_parent(pending_view_child->widget, GTK_WIDGET(view));
+      gtk_widget_set_parent(pending_child, GTK_WIDGET(self));
     }
   }
 
-  for (GList* child = view->children_list; child; child = child->next) {
-    FlViewChild* view_child = reinterpret_cast<FlViewChild*>(child->data);
-    if (view_child) {
+  for (GtkWidget* child : *self->children_list) {
+    if (child) {
       // removed child
-      g_object_ref(view_child->widget);
-      gtk_widget_unparent(view_child->widget);
-      g_free(view_child);
-      child->data = nullptr;
+      g_object_ref(child);
+      gtk_widget_unparent(child);
     }
   }
 
-  g_list_free(view->children_list);
-  view->children_list = view->pending_children_list;
-  view->pending_children_list = nullptr;
+  self->children_list->clear();
+  std::swap(self->children_list, self->pending_children_list);
+
+  for (auto& it : *self->last_clipping_views) {
+    g_object_unref(it.second);
+  }
+  self->last_clipping_views->clear();
+  std::swap(self->last_clipping_views, self->current_clipping_views);
 
   struct _ReorderData data = {
       .parent_window = gtk_widget_get_window(GTK_WIDGET(view)),
```

Comment:
this should probably be `gtk_widget_get_window(GTK_WIDGET(self))`

Reply:
Done.